File: examples/plugins/c/truncation/weight_truncator.c
// Example propaq truncator plugin, implemented in C.
//
// Weight alone, so it declares no dependencies
//

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define PROPAQ_TRUNCATOR_ABI_VERSION 1u

typedef struct {
    uint32_t max_weight;
} Ctx;

uint32_t propaq_truncator_abi_version(void) {
    return PROPAQ_TRUNCATOR_ABI_VERSION;
}

static uint32_t parse_max_weight(const char* config_json) {
    if (config_json == NULL) return UINT32_MAX;
    const char* key = strstr(config_json, "\"max_weight\"");
    if (key == NULL) return UINT32_MAX;
    const char* colon = strchr(key, ':');
    if (colon == NULL) return UINT32_MAX;
    return (uint32_t)strtoul(colon + 1, NULL, 10);
}

void* propaq_truncator_create(const char* config_json) {
    Ctx* ctx = (Ctx*)malloc(sizeof(Ctx));
    ctx->max_weight = parse_max_weight(config_json);
    return ctx;
}

void propaq_truncator_destroy(void* ctx) {
    free(ctx);
}

int32_t propaq_truncator_keep(void* ctx, uint32_t basis_kind, const uint64_t* words, size_t n_words,
                              uint32_t n_units, uint32_t weight, double coeff_magnitude,
                              uint32_t layer_index, uint32_t n_layers) {
    (void)basis_kind; (void)words; (void)n_words; (void)n_units;
    (void)coeff_magnitude; (void)layer_index; (void)n_layers;
    return weight <= ((Ctx*)ctx)->max_weight ? 1 : 0;
}

int32_t propaq_truncator_keep_batch(void* ctx, uint32_t basis_kind, const uint64_t* words,
                                    size_t n_words_per_term, uint32_t n_units,
                                    const uint32_t* weights, const double* coeff_magnitudes,
                                    uint32_t layer_index, uint32_t n_layers, uint8_t* out_keep,
                                    size_t n_terms) {
    (void)basis_kind; (void)words; (void)n_words_per_term; (void)n_units;
    (void)coeff_magnitudes; (void)layer_index; (void)n_layers;
    uint32_t max_weight = ((Ctx*)ctx)->max_weight;
    for (size_t i = 0; i < n_terms; i++) {
        out_keep[i] = weights[i] <= max_weight ? 1 : 0;
    }
    return 0;
}

File: examples/plugins/julia/truncation/weight_truncator.jl
# Example propaq truncator plugin, implemented in Julia (AOT-compiled via
# PackageCompiler.jl into a C-ABI shared library; see the plugins README).
#
# Weight alone, so it declares no dependencies
#

const PROPAQ_TRUNCATOR_ABI_VERSION = UInt32(1)

mutable struct Ctx
    max_weight::UInt32
end

const CTX_STORE = Dict{Ptr{Cvoid},Ctx}()
const CTX_LOCK = ReentrantLock()

function parse_field(json::AbstractString, key::AbstractString, fallback::Float64)
    m = match(Regex("\"" * key * "\"\\s*:\\s*(-?[0-9.eE+-]+)"), json)
    m === nothing ? fallback : parse(Float64, m.captures[1])
end

Base.@ccallable function propaq_truncator_abi_version()::UInt32
    PROPAQ_TRUNCATOR_ABI_VERSION
end

Base.@ccallable function propaq_truncator_create(config_json::Ptr{Cchar})::Ptr{Cvoid}
    json = config_json == C_NULL ? "" : unsafe_string(config_json)
    ctx = Ctx(UInt32(parse_field(json, "max_weight", 4.0)))
    key = pointer_from_objref(ctx)
    lock(CTX_LOCK) do
        CTX_STORE[key] = ctx
    end
    return key
end

Base.@ccallable function propaq_truncator_destroy(ctx::Ptr{Cvoid})::Cvoid
    lock(CTX_LOCK) do
        delete!(CTX_STORE, ctx)
    end
    return nothing
end

Base.@ccallable function propaq_truncator_keep(ctx::Ptr{Cvoid}, basis_kind::UInt32,
                                               words::Ptr{UInt64}, n_words::Csize_t,
                                               n_units::UInt32, weight::UInt32,
                                               coeff_magnitude::Cdouble,
                                               layer_index::UInt32, n_layers::UInt32)::Int32
    c = unsafe_pointer_to_objref(ctx)::Ctx
    return weight <= c.max_weight ? Int32(1) : Int32(0)
end

Base.@ccallable function propaq_truncator_keep_batch(ctx::Ptr{Cvoid}, basis_kind::UInt32,
                                                     words::Ptr{UInt64}, n_words_per_term::Csize_t,
                                                     n_units::UInt32, weights::Ptr{UInt32},
                                                     coeff_magnitudes::Ptr{Cdouble},
                                                     layer_index::UInt32, n_layers::UInt32,
                                                     out_keep::Ptr{UInt8}, n_terms::Csize_t)::Int32
    c = unsafe_pointer_to_objref(ctx)::Ctx
    w = unsafe_wrap(Array, weights, n_terms)
    result = unsafe_wrap(Array, out_keep, n_terms)
    @inbounds for i in 1:n_terms
        result[i] = w[i] <= c.max_weight ? UInt8(1) : UInt8(0)
    end
    return Int32(0)
end
